### ISSUE TITLE
What is a Neural Network?

### ISSUE THREAD

**Issue by guevara** [guevara/read-it-later#5734]
What is a Neural Network?<br>
<br>
https://ift.tt/2H6o6a5<br>
<br>
About 0 Minutes<br>
<br>
<div>
<div>
<p>Neural networks are important! They diagnose illnesses, help you find photos of your cat, decide whether to give you a loan<sup><a href="https://norasandler.com/2017/10/20/What-is-a-Neural-Network.html#fn1">1</a></sup>. They make up a huge part of what we call &ldquo;machine learning&rdquo; or &ldquo;artificial intelligence,&rdquo; especially the new, exciting, scary parts of it. And, given the magnitude of the problems they can handle, they&rsquo;re simpler than you&rsquo;d expect.</p>
<p>Most explanations of neural networks for lay people are kind of lousy, because they always start by talking about brains. They&rsquo;ll define neural networks as &ldquo;computer systems modeled on the human brain&rdquo; or &ldquo;programs that learn the same way as people&rdquo; or even as &ldquo;artificial brains.&rdquo; This makes neural networks sound really cool and sexy and also like they might become sentient and destroy us at any moment. But it doesn&rsquo;t give you any intuition for what they actually do, because neural networks aren&rsquo;t really modeled on brains. They&rsquo;re loosely inspired by brains, the same way Hollywood movies are &ldquo;inspired by true events.&rdquo; I really dislike the comparison to brains, because it creates a completely unnecessary air of mystery around a system that&rsquo;s basically just a lot of math. So I wanted to write an introduction to neural networks that demystifies them a bit.</p>
<p>This blog post explains what a neural network is, in a way that assumes no math background and contains no brain metaphors. Without all the math, it won&rsquo;t be a very complete or very precise explanation, but I hope it will still give you a decent sense of what neural networks do. At each step, I&rsquo;ll fill in some of the math-y details in a separate section; feel free to read it or skip it.</p>
<h1>Defining the Task</h1>
<p>In a typical machine learning task, you&rsquo;re trying to find a mathematical formula, or model, that takes in some information and uses it to answer a specific question. For example, you might train a model to:</p>
<p>In this context, the machine isn&rsquo;t learning to speak or play chess or play the violin, it&rsquo;s just &ldquo;learning&rdquo; a formula that you plug numbers into and get numbers out of. The hard part, of course, is figuring out exactly what formula to use - this is called training the model. This blog post focuses on supervised machine learning, where we train the model by feeding it a bunch of examples that have already been labelled with the correct answer. (You can also use neural networks for unsupervised learning, which uses unlabelled data, but that&rsquo;s beyond the scope of this blog post.)</p>
<p>You can think of training a model as trying to come up with the best formula to predict something. You can also think of it as finding the line of best fit for your data:</p>
<p><img alt="Random data points on a graph and their line of best fit." src="https://norasandler.com/assets/best_fit.svg"></p>
<h1>Training the Model</h1>
<p>One common technique for training a supervised learning model is an algorithm called <strong>gradient descent</strong>. Here&rsquo;s the gist of it:</p>
<ol>
<li>Get some sample data that includes both the input and the answer you want for the question you care about.</li>
<li>Make an initial guess about what formula to use.</li>
<li>Run your formula on all your training data to get a bunch of predicted answers.</li>
<li>Calculate how close your predicted answers were to the real answers - the <strong>error</strong>.</li>
<li>Make some tiny adjustments to your formula to reduce the error a bit.</li>
<li>Repeat steps 3-5 until the error is pretty small.</li>
</ol>
<p>Say you&rsquo;re throwing a dinner party and you need to know how much wine to buy. You could just consult Emily Post, but instead, you decide to use MACHINE LEARNING. Let&rsquo;s go through the steps above:</p>
<ol>
<li>
<p>You need training data. So you call three friends who just threw really great dinner parties, and you ask them how many people they invited and how much wine they bought.</p>
<table>
<thead>
<tr>
<th>Number of Guests</th>
<th>Bottles of wine consumed</th>
</tr>
</thead>
<tbody>
<tr>
<td>3</td>
<td>2</td>
</tr>
<tr>
<td>10</td>
<td>7</td>
</tr>
<tr>
<td>75</td>
<td>50</td>
</tr>
</tbody>
</table>
</li>
<li>
<p>You assume that on average, each person will drink about the same amount of wine, so you can model how much to buy with the following formula:</p>
<p>Our goal is to find the best value of A. You make an initial guess of one bottle of wine per person, so you set A=1, which gives you:</p>
</li>
<li>
<p>For each dinner party in your dataset, you use your formula to calculate how much wine to buy:</p>
<table>
<thead>
<tr>
<th>Number of guests</th>
<th>Predicted bottles of wine</th>
</tr>
</thead>
<tbody>
<tr>
<td>3</td>
<td>3</td>
</tr>
<tr>
<td>10</td>
<td>10</td>
</tr>
<tr>
<td>75</td>
<td>75</td>
</tr>
</tbody>
</table>
</li>
<li>
<p>Now compare your prediction to the correct answers from your training set, and see how close you were:</p>
<table>
<thead>
<tr>
<th>Number of guests</th>
<th>Predicted bottles of wine</th>
<th>Actual bottles of wine</th>
<th>Error</th>
</tr>
</thead>
<tbody>
<tr>
<td>3</td>
<td>3</td>
<td>2</td>
<td>1</td>
</tr>
<tr>
<td>10</td>
<td>10</td>
<td>7</td>
<td>3</td>
</tr>
<tr>
<td>75</td>
<td>75</td>
<td>50</td>
<td>25</td>
</tr>
</tbody>
</table>
</li>
<li>
<p>This formula overestimated how much wine you need, so you reduce A to 0.8. The new formula is:</p>
</li>
<li>
<p>You try again with your new formula:</p>
<table>
<thead>
<tr>
<th>Number of guests</th>
<th>Predicted bottles of wine</th>
<th>Actual bottles of wine</th>
<th>Error</th>
</tr>
</thead>
<tbody>
<tr>
<td>3</td>
<td>2.4</td>
<td>2</td>
<td>0.4</td>
</tr>
<tr>
<td>10</td>
<td>8</td>
<td>7</td>
<td>1</td>
</tr>
<tr>
<td>75</td>
<td>60</td>
<td>50</td>
<td>10</td>
</tr>
</tbody>
</table>
<p>That&rsquo;s better, but you&rsquo;re still overestimating how much wine to buy, so you reduce A a bit more. As you get closer and closer to the right value for A, you&rsquo;ll make smaller and smaller adjustments, so you don&rsquo;t overshoot the best answer. You&rsquo;ll continue this process until you have a pretty good model, which in this case will be around A=0.67. Then you can figure out how much wine you need and throw a fabulous party. &#127863;</p>
</li>
</ol>
<p>In this example we sort of eyeballed how to adjust A. Real machine learning programs will automatically calculate the error and adjust the model parameters, repeating this process hundreds or thousands of times.</p>
<div>
<h1>Math-y Details</h1>
<p>In a more realistic problem, you would also want a constant term in the model: Most machine learning problems use more than one input (e.g. to predict housing prices you&rsquo;d look at square feet and number of bedrooms and zip code and so on). So you can generalize this to:</p>
<p>Each here is called an <strong>input feature</strong>.</p>
<p>In step 4) we calculated the error for each data point, but you actually need to measure how well the model does overall, across the whole data set. You can measure this using the mean squared error, or L2 loss.</p>
<p>Some notation: if there are data points in the training set, I&rsquo;ll write the correct answer for the data point as . (The i there isn&rsquo;t an exponent - I&rsquo;m using a superscript to label different data points because I&rsquo;m using subscripts to label different input features for the same data point.) I&rsquo;ll use to notate the answer the model predicts for the ith data point. So the equation for the overall error (also called the cost function) will be:</p>
<p>E.g. with the data from step 4:</p>
<p>Step 5 is updating each parameter in a way that will reduce this error. In the simplest case, where there&rsquo;s just one input (e.g. number of guests) and just one parameter , you can use this formula to adjust that parameter (where the left arrow means we&rsquo;re assigning a new value to a):</p>
<p>( is the <strong>learning rate</strong> - it controls how large an adjustment to make at each step. Unlike the other values here, which are calculated automatically, you have to choose yourself.)</p>
<p>To understand why this works, imagine plotting the error for different values of parameter : <img alt='Hand-drawn, sort of parabola-shaped graph where x-axis is a (the parameter) and y-axis is error. Minimum of the graph is labelled "best value of a."' src="https://norasandler.com/assets/cost_graph.svg"> You&rsquo;re looking for the value of that makes the error smallest - the minimum of this graph. At the minimum, the slope of the graph will be 0, so . If is too small, this derivative will be negative, so the formula above will increase it. If is too big, the derivative will be positive, so the formula will decrease . The further away is from the minimum, the steeper the slope, so the larger the adjustment to .</p>
<p>Note that this process will only find a local minimum, not necessarily a global minimum.</p>
<p><img alt="Hand-drawn, squiggly graph where x-axis is a (the parameter) and y-axis is error. There are multiple local minima, labelled with confused question marks" src="https://norasandler.com/assets/bad_cost_graph.svg"></p>
<p>But this particular error function will only have one local minimum, so it works out.</p>
<p>To generalize this to multiple parameters, just take the partial derivative of the error with respect to each of those parameters:</p>
</div>
<h1>Neural Networks vs. Other Models</h1>
<p>The example above used a linear model: for every extra person you invite, the amount of wine you need goes up by a fixed amount. Depending on the parameters you find during gradient descent, your linear model might look like any of these<sup><a href="https://norasandler.com/2017/10/20/What-is-a-Neural-Network.html#fn2">2</a></sup>:</p>
<p>But no matter what parameters you use, it will always be a straight line.</p>
<p>But a linear equation might not always fit your data well. The best model for your data might be quadratic:</p>
<p><img alt="Plot of speed vs stopping distance with a quadratic fit" src="https://norasandler.com/assets/quadratic.png"></p>
<p>Or logarithmic<sup><a href="https://norasandler.com/2017/10/20/What-is-a-Neural-Network.html#fn3">3</a></sup>:</p>
<p><img alt="Graph of logarithm" src="https://norasandler.com/assets/log.png"></p>
<p>&hellip;or something else entirely. The point is, this approach works well <strong>if you can choose an appropriate model</strong>, one that&rsquo;s basically the correct shape. But to choose an appropriate model, you already need to understand how your inputs relate to the correct answer. You can sort of eyeball this on a graph if you have one input, but not if you have a hundred.</p>
<p>Let&rsquo;s go back to the dinner party example. Maybe you didn&rsquo;t get nice linear data - maybe you can&rsquo;t see any correlation between the size of a party and how much wine was consumed. So you call your friends back and ask them about their parties. It turns out that a lot of factors go into dinner party planning. Of course, some people drink and others don&rsquo;t. But there are other complications. Your Aunt Agatha glares disapprovingly at drinkers; if she is invited, everyone will stick to ginger ale. Your Aunt Dahlia is the life of the party; if she&rsquo;s invited, everyone will have more than usual. (The question of why your aunts are invited to your friends&rsquo; dinner parties is beyond the scope of this blog post.) Last year, two of your friends had a nasty divorce and everyone took sides; if people from both factions come to the party, at least one faction will leave in a huff before you&rsquo;ve served the appetizers. And that&rsquo;s just the tip of the iceberg - there may be dozens or hundreds of similar factors at play that you don&rsquo;t even know about.</p>
<p>This is the sort of problem neural networks are good at. They&rsquo;re flexible enough to fit any relationship between your inputs and output, so you don&rsquo;t need to understand how they relate ahead of time. And they can figure out how different combinations of inputs will impact your output. A particular guest at your dinner party will have different impacts on different groups of people - a neural network can suss out the different rules for all these different interactions.</p>
<p>The math behind them isn&rsquo;t complicated - it&rsquo;s just function composition, calling functions on the outputs of other functions. So, where the example above looked like this:</p>
<p align="center">$$ \textrm{Answer} = \textrm{MATH}(x_1, x_2, \cdots, x_n) $$</p>
<p><img alt="Diagram with three boxes labelled DATA leading to a box labelled MATH leading to a box labelled ANSWER" src="https://norasandler.com/assets/linear_regression_graph.png"></p>
<p>Then a neural net is just this:</p>
<p align="center">$$ \textrm{Answer} = \textrm{MATH}(\textrm{MATH}(x_1, x_2, \cdots, x_n) + \textrm{MATH}(x_1, x_2, \cdots, x_n) + \cdots) $$</p>
<p><img alt='Diagram with three boxes labelled DATA on the left, three boxes labelled MATH in the middle, another box labelled "even more math" to the right, and a box labelled ANSWER furthest right. Each DATA box leads to all three MATH boxes, each MATH box leads into the "even more math" box which leads into the answer' src="https://norasandler.com/assets/neural_net_graph.png"></p>
<p>Each of the &ldquo;math&rdquo; boxes above is called a <strong>neuron</strong>. The idea is that each neuron in the first layer (the orange column in the figure) will calculate some relevant information, like &ldquo;are people from Friend Group A and Friend Group B present?&rdquo; directly from the inputs you give it. Each bit of relevant information is a <strong>feature</strong>. Then the neuron in the last layer uses all these features to calculate the answer you really care about. You don&rsquo;t know in advance what features are important; the neural network will figure them out when you train it.</p>
<p>Neural networks work because individual neurons can capture information about the interactions between inputs, instead of trying to understand each input in isolation. In many machine learning problems, you can&rsquo;t learn a whole lot from looking at one input at a time; you have to consider how your different inputs interact. At your dinner party, the presence or absence of a single guest can only tell you so much; the social dynamics between guests dictate how much wine they drink. Or think of a photograph: you can&rsquo;t learn anything from the color of a single pixel. You have to look at the contrast between pixels to make out edges, shapes, and objects.</p>
<p>Let&rsquo;s build a neural network for our dinner party example. Suppose you have three people you could invite to a dinner party: your aunts Agatha and Dahlia, and your best friend Stiffy. Since just the number of guests doesn&rsquo;t give you enough information, you need to use the whole guest list as an input. Let&rsquo;s represent it as a list of numbers, where each position in the list represents one possible guest, in this order:</p>
<p>A zero means the friend is not at the party, and a one means they are at the party. So if only Agatha is coming, the list will be:</p>
<p>And if Dahlia and Stiffy are coming, it will be:</p>
<p>Now you need to figure out what the first layer of neurons should calculate. You&rsquo;ll need a function called ReLU (short for rectified linear unit), which is almost identical to the linear function from the earlier example, except it flattens out at zero if its input is ever negative:</p>
<p><img alt="Graph of ReLU function" src="https://norasandler.com/assets/relu.png"></p>
<p>Here&rsquo;s a slightly more precise version of the neural network diagram above:</p>
<p><img alt='Diagram with three boxes labelled Guest on the left, three orange boxes labelled Neuron 1, Neuron 2, and Neuron 3 in the middle, another box labelled "Neuron 4" to the right, and a box labelled Number of bottles furthest right. Each DATA box leads to each of neuron 1, neuron 2, and neuron 3, and each of these leads into "neuron 4" box which leads to the "bottles of wine" box' src="https://norasandler.com/assets/neural_net_graph_2.png"></p>
<p>Based on this diagram, you have two layers in your neural network. The orange boxes make up layer one. We&rsquo;ll label the outputs of the three neurons in this layer , , and . Each neuron will have its own set of parameters , , , and . You&rsquo;ll need to figure out the values of these parameters when you train the model. Then each neuron will calculate:</p>
<p>So, for example, if only Stiffy is coming to your party, would calculate:</p>
<p>The second layer of the network has only one neuron - in the diagram above. This neuron will have three more parameters , , , and , which you&rsquo;ll also need to find by training the model. It will calculate:</p>
<p>If you put it all together, the formula will be:</p>
<p>Now, just like in the last section, you need to figure out your parameters: , , , , , and so on. It turns out that you can use exactly the same system as when you were just looking at number of guests; the only difference is that you&rsquo;ll first do the &ldquo;tiny adjustments&rdquo; step for the last layer, then work your way backwards to first layer. This system of making corrections one layer at a time is called <strong>backpropagation of errors</strong>.</p>
<p>So the new training process is:</p>
<ol>
<li>
<p>Convince your friends to give you the guest lists for all their parties.</p>
</li>
<li>
<p>Choose random values for every parameter to create an initial model.</p>
</li>
<li>
<p>Run all the guest lists through the model.</p>
</li>
<li>
<p>See whether the model over- or under-estimated how much wine was needed for each guest list.</p>
</li>
<li>
<p>Adjust , , , and by a little to reduce the error. This means you need to figure out whether each individual neuron tends to overestimate or underestimate. For example, if the model tends to overestimate when the first neuron spits out a big value, you need to make a bit smaller; but if you mostly overestimate when the second neuron is big, then you need to decrease more than .</p>
</li>
<li>
<p>Adjust , , &hellip;, , by a little to reduce the error. You can re-use the information from step 5 about when individual neurons over- or under-estimate. You might find that tends to overestimate when Agatha is invited to a party, but underestimate when Dahlia is invited; then you&rsquo;d want to decrease and increase . That will help correct , which will make the final answer a little more accurate.</p>
</li>
<li>
<p>Repeat steps 3 - 6 until the error is pretty small. Like in the last example, your program will automatically run the model, calculate the error, and adjust all the parameters hundreds or thousands of times.</p>
</li>
</ol>
<p>When you&rsquo;re finally done training the model, you might end up with something close to this:</p>
<p>In this model, each neuron encodes something about people&rsquo;s drinking habits:</p>
<ul>
<li>Neuron 1 mean that, left to her own devices, Stiffy will have about half a bottle of wine.</li>
<li>Neuron 2 means that when Dahlia is around, she&rsquo;ll have a couple bottles, and Stiffy will have more than usual.</li>
<li>Neuron 3 means that when Agatha is present, nobody drinks anything.</li>
<li>And that at the end, our value, means that if nobody shows up you&rsquo;ll have a bottle of wine by yourself.</li>
</ul>
<p>And that, more or less, is how neural networks can help you plan a dinner party.</p>
<p><em>If you notice any mistakes in this blog post, please <a href="https://norasandler.com/cdn-cgi/l/email-protection#402e2f3221002e2f322133212e242c25326e232f2d">email me</a>! This post mostly rehashes the <a href="https://www.coursera.org/learn/neural-networks-deep-learning">Coursera course</a> on neural networks and deep learning, so if you enjoyed this post you might want to take the course.</em></p>
<p><em>In an earlier version of this post, the error in the &ldquo;Math-y Details&rdquo; section was calculated incorrectly. Thanks to KGruel for the correction!</em></p>
</div>
</div>
<br>
<br>
<br>
<br>
<br>
<br>
via norasandler.com https://norasandler.com<br>
<br>
November 8, 2019 at 02:36PM
